File: NewsApp/app/src/main/java/com/example/shakil/newsapp/MainActivity.java
package com.example.shakil.newsapp;

import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import android.app.AlertDialog;
import android.os.Bundle;

import com.example.shakil.newsapp.Adapter.ListSourceAdapter;
import com.example.shakil.newsapp.Common.Common;
import com.example.shakil.newsapp.Interface.NewsService;
import com.example.shakil.newsapp.Model.WebSite;
import com.google.gson.Gson;

import dmax.dialog.SpotsDialog;
import io.paperdb.Paper;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class MainActivity extends AppCompatActivity {

    RecyclerView listWebSite;
    RecyclerView.LayoutManager layoutManager;
    NewsService mService;
    ListSourceAdapter adapter;
    AlertDialog dialog;
    SwipeRefreshLayout swipeLayout;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        /*Paper init*/
        Paper.init(this);

        /*Init Service*/
        mService = Common.getNewsService();

        /*Init View*/
        swipeLayout = findViewById(R.id.swipeRefresh);
        swipeLayout.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {
                loadWebsiteService(true);
            }
        });

        listWebSite = findViewById(R.id.list_source);
        listWebSite.setHasFixedSize(true);
        layoutManager = new LinearLayoutManager(this);
        listWebSite.setLayoutManager(layoutManager);

        dialog = new SpotsDialog.Builder().setContext(this).setCancelable(false).build();

        loadWebsiteService(false);
    }

    private void loadWebsiteService(boolean isRefreshed) {
        if (!isRefreshed){
            String cache = Paper.book().read("cache");

            /*If have cache*/
            if (cache != null && !cache.isEmpty() && !cache.equals("null")) {
                /*Convert cache from Json to Object*/
                WebSite webSite = new Gson().fromJson(cache, WebSite.class);

                adapter = new ListSourceAdapter(getBaseContext(), webSite);
                adapter.notifyDataSetChanged();
                listWebSite.setAdapter(adapter);
            }
            /*If not have cache*/
            else {
                dialog.show();

                /*Fetch new Data*/
                mService.getSources().enqueue(new Callback<WebSite>() {
                    @Override
                    public void onResponse(Call<WebSite> call, Response<WebSite> response) {
                        adapter = new ListSourceAdapter(getBaseContext(), response.body());
                        adapter.notifyDataSetChanged();
                        listWebSite.setAdapter(adapter);

                        /*Save to cache*/
                        Paper.book().write("cache", new Gson().toJson(response.body()));

                        dialog.dismiss();
                    }

                    @Override
                    public void onFailure(Call<WebSite> call, Throwable t) {

                    }
                });
            }
        }
        /*If from Swipe to refresh*/
        else {
            swipeLayout.setRefreshing(true);
            /*Fetch new Data*/
            mService.getSources().enqueue(new Callback<WebSite>() {
                @Override
                public void onResponse(Call<WebSite> call, Response<WebSite> response) {
                    adapter = new ListSourceAdapter(getBaseContext(), response.body());
                    adapter.notifyDataSetChanged();
                    listWebSite.setAdapter(adapter);

                    /*Save to cache*/
                    Paper.book().write("cache", new Gson().toJson(response.body()));

                    swipeLayout.setRefreshing(false);
                }

                @Override
                public void onFailure(Call<WebSite> call, Throwable t) {

                }
            });
        }
    }
}

File: KotlinNewsApp/app/src/main/java/com/example/shakil/kotlinnewsapp/MainActivity.kt
package com.example.shakil.kotlinnewsapp

import android.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.widget.Toast
import androidx.recyclerview.widget.LinearLayoutManager
import com.example.shakil.kotlinnewsapp.Adapter.ListSourceAdapter
import com.example.shakil.kotlinnewsapp.Common.Common
import com.example.shakil.kotlinnewsapp.Interface.NewsService
import com.example.shakil.kotlinnewsapp.Model.WebSite
import com.google.gson.Gson
import dmax.dialog.SpotsDialog
import io.paperdb.Paper
import kotlinx.android.synthetic.main.activity_main.*
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class MainActivity : AppCompatActivity() {

    lateinit var layoutManager: LinearLayoutManager
    lateinit var mService: NewsService
    lateinit var adapter: ListSourceAdapter
    lateinit var dialog: AlertDialog

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        /*Init cache DB*/
        Paper.init(this)

        /*Init Service*/
        mService = Common.newsService

        /*Init View*/
        swipe_to_refresh.setOnRefreshListener {
            loadWebSiteService(true)
        }

        recycler_view_source_news.setHasFixedSize(true)
        layoutManager = LinearLayoutManager(this)
        recycler_view_source_news.layoutManager = layoutManager

        dialog = SpotsDialog.Builder().setContext(this).setCancelable(false).build()

        loadWebSiteService(false)
    }

    private fun loadWebSiteService(isRefresh: Boolean) {
        if (!isRefresh) {
            val cache = Paper.book().read<String>("cache")
            if (cache != null && !cache.isBlank() && cache != null) {
                /*Read cache*/
                val webSite = Gson().fromJson<WebSite>(cache, WebSite::class.java)
                adapter = ListSourceAdapter(baseContext, webSite)
                adapter.notifyDataSetChanged()
                recycler_view_source_news.adapter = adapter
            }
            else {
                /*Load website and write cache*/
                dialog.show()

                /*Fetch new Data*/
                mService.sources.enqueue(object : Callback<WebSite> {
                    override fun onFailure(call: Call<WebSite>, t: Throwable) {
                        Toast.makeText(baseContext, "Failed", Toast.LENGTH_SHORT).show()
                    }

                    override fun onResponse(call: Call<WebSite>, response: Response<WebSite>) {
                        adapter = ListSourceAdapter(baseContext, response.body()!!)
                        adapter.notifyDataSetChanged()
                        recycler_view_source_news.adapter = adapter

                        /*Save to cache*/
                        Paper.book().write("cache", Gson().toJson(response.body()))

                        dialog.dismiss()
                    }

                })
            }
        }
        else {
            swipe_to_refresh.isRefreshing = true
            /*Fetch new Data*/
            mService.sources.enqueue(object : Callback<WebSite> {
                override fun onFailure(call: Call<WebSite>, t: Throwable) {
                    Toast.makeText(baseContext, "Failed", Toast.LENGTH_SHORT).show()
                }

                override fun onResponse(call: Call<WebSite>, response: Response<WebSite>) {
                    adapter = ListSourceAdapter(baseContext, response.body()!!)
                    adapter.notifyDataSetChanged()
                    recycler_view_source_news.adapter = adapter

                    /*Save to cache*/
                    Paper.book().write("cache", Gson().toJson(response.body()))

                    swipe_to_refresh.isRefreshing = false
                }

            })
        }
    }
}
